In [1]:
import ollama

print(ollama.list())

models=[Model(model='qwen2.5:3b', modified_at=datetime.datetime(2026, 9, 22, 13, 20, 45, 949888, tzinfo=TzInfo(19800)), digest='357c53fb659c5076de1d65ccb0b397446227b71a42be9d1603d46168015c9e4b', size=1929912432, details=ModelDetails(parent_model='', format='gguf', family='qwen2', families=['qwen2'], parameter_size='3.1B', quantization_level='Q4_K_M')), Model(model='llama3.2:3b', modified_at=datetime.datetime(2026, 9, 22, 12, 46, 36, 193685, tzinfo=TzInfo(19800)), digest='a80c4f17acd55265feec403c7aef86be0c25983ab279d83f3bcd3abbcb5b8b72', size=2019393189, details=ModelDetails(parent_model='', format='gguf', family='llama', families=['llama'], parameter_size='3.2B', quantization_level='Q4_K_M'))]


In [2]:
import ollama

question = "Explain how a PID controller works in robotics."

system_prompts = {
    "Basic": "You are a helpful assistant.",
    
    "Role-based": (
        "You are a robotics engineering professor. "
        "Explain concepts accurately to an undergraduate engineering student."
    ),
    
    "Structured": (
        "You are a robotics expert. "
        "Explain the concept clearly using: Definition, Working, "
        "Example, and Key Points."
    )
}

for name, system_prompt in system_prompts.items():
    response = ollama.chat(
        model="llama3.2:3b",
        messages=[
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": question}
        ]
    )
    
    print(f"\n===== {name} System Prompt =====")
    print(response["message"]["content"])


===== Basic System Prompt =====
A PID (Proportional-Integral-Derivative) controller is a type of feedback control system commonly used in robotics to regulate and stabilize systems. Here's a detailed explanation of how a PID controller works:

**Basic Components**

A PID controller consists of three main components:

1. **Sensor**: Measures the error between the desired setpoint (target value) and the current system output (e.g., joint angle, position, or velocity).
2. **Controller**: Processes the sensor data and calculates the control output.
3. **Actuator**: Applies the control output to the system (e.g., motor, servo, or valve).

**PID Controller Operation**

The PID controller operates as follows:

1. **Proportional (P)**: The controller calculates the proportional error, which is the difference between the desired setpoint and the current system output. The proportional term is proportional to the error and is used to adjust the control output in the same direction as the error.

In [3]:
import ollama

question = "Explain how a PID controller works in robotics."

system_prompts = {
    "Basic": "You are a helpful assistant. Answer in 40 words or less.",
    
    "Role-based": (
        "You are a robotics engineering professor. "
        "Explain accurately to an undergraduate student in 40 words or less."
    ),
    
    "Structured": (
        "You are a robotics expert. "
        "Answer using Definition, Working, and Example. "
        "Keep the complete answer under 40 words."
    )
}

for name, system_prompt in system_prompts.items():
    response = ollama.chat(
        model="llama3.2:3b",
        messages=[
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": question}
        ]
    )

    answer = response["message"]["content"].replace("\n", " ")
    
    print(f"\n===== {name} System Prompt =====")
    print(answer)


===== Basic System Prompt =====
A PID (Proportional-Integral-Derivative) controller in robotics uses feedback to regulate and control the system's behavior. It compares the desired output to the current output, adjusting the control input to minimize error, with proportional, integral, and derivative terms.

===== Role-based System Prompt =====
A PID (Proportional-Integral-Derivative) controller in robotics adjusts the motor speed or power to achieve a desired position, velocity, or acceleration. It calculates the error, integrates it, and adjusts the output to minimize the error over time.

===== Structured System Prompt =====
**Definition:** PID (Proportional-Integral-Derivative) controller regulates a process by adjusting an input to achieve a desired output, maintaining stability and accuracy.  **Working:** PID controller calculates the error between desired and actual values, adjusting the input based on proportional, integral, and derivative terms.  **Example:** A PID controller

In [4]:
import ollama

few_shot_prompt = """
You are a robotics tutor.

Example 1:
Question: What is LiDAR?
Answer: LiDAR uses laser pulses to measure distances and create a representation of the environment.

Example 2:
Question: What is a camera sensor?
Answer: A camera captures visual information that robots can use for perception and object recognition.

Now answer:
Question: What is an ultrasonic sensor?
"""

response = ollama.chat(
    model="llama3.2:3b",
    messages=[
        {"role": "user", "content": few_shot_prompt}
    ]
)

print(response["message"]["content"])

Question: What is an ultrasonic sensor?
Answer: An ultrasonic sensor uses high-frequency sound waves to measure distances and detect objects around the robot, typically measuring up to 10 meters or more in range.


In [5]:
import ollama

question = "Explain how an autonomous robot avoids obstacles."

zero_shot = ollama.chat(
    model="llama3.2:3b",
    messages=[
        {
            "role": "user",
            "content": question
        }
    ]
)

few_shot = ollama.chat(
    model="llama3.2:3b",
    messages=[
        {
            "role": "user",
            "content": """
You are a robotics tutor.

Example:
Question: What is LiDAR?
Answer: LiDAR measures distances using laser pulses and helps robots understand their surroundings.

Example:
Question: What is ultrasonic sensing?
Answer: Ultrasonic sensing uses sound waves to measure distance and detect nearby objects.

Now answer:
Question: Explain how an autonomous robot avoids obstacles.
"""
        }
    ]
)

print("===== ZERO-SHOT =====")
print(zero_shot["message"]["content"])

print("\n===== FEW-SHOT =====")
print(few_shot["message"]["content"])

===== ZERO-SHOT =====
An autonomous robot uses a combination of sensors, software, and algorithms to avoid obstacles in its environment. Here's a general overview of the steps involved in obstacle avoidance:

1. **Sensor Suite**: The robot is equipped with a range of sensors that provide information about its surroundings, such as:
	* LIDAR (Light Detection and Ranging): Creates a 3D map of the environment.
	* Camera: Provides visual information about the environment.
	* Ultrasonic sensors: Detect obstacles at close range.
	* Inertial Measurement Unit (IMU): Measures the robot's acceleration, roll, pitch, and yaw.
2. **Data Processing**: The sensor data is processed by the robot's onboard computer, which uses algorithms to analyze the data and make decisions about obstacle avoidance.
3. **Object Detection**: The robot uses object detection algorithms to identify obstacles in its path. This can include techniques such as:
	* Edge detection: Identifying the edges of objects.
	* Feature d

### Zero-Shot vs Few-Shot Prompting

Zero-shot prompting asks the model to answer a question without providing examples. In this experiment, the zero-shot response was detailed and included multiple sensors, algorithms, and obstacle-avoidance techniques.

Few-shot prompting provides examples before the actual question. The few-shot response followed the demonstrated structure and focused on common sensor-based obstacle avoidance methods.

**Observation:** Few-shot prompting can guide the model toward a more consistent response style, while zero-shot prompting allows the model to generate an answer without example-based guidance.

In [6]:
import ollama

question = "Explain LiDAR in robotics."

formats = {
    "Short": "Answer in exactly 2 sentences.",
    "Bullet": "Answer using exactly 3 bullet points.",
    "Technical": "Answer in 3 sentences and include one technical term."
}

for name, instruction in formats.items():
    response = ollama.chat(
        model="llama3.2:3b",
        messages=[
            {
                "role": "system",
                "content": f"You are a robotics tutor. {instruction}"
            },
            {
                "role": "user",
                "content": question
            }
        ]
    )

    print(f"\n===== {name.upper()} FORMAT =====")
    print(response["message"]["content"])


===== SHORT FORMAT =====
LiDAR (Light Detection and Ranging) in robotics uses laser light to create high-resolution 3D point clouds of the environment, allowing robots to perceive and navigate their surroundings with precision and accuracy. By emitting pulses of laser light and measuring the time-of-flight and wavelength of the reflected light, LiDAR systems can build detailed maps of their environment, enabling robots to avoid obstacles, track objects, and perform complex tasks.

===== BULLET FORMAT =====
Here are three key points about LiDAR in robotics:

• **Definition and Function**: LiDAR (Light Detection and Ranging) is a remote sensing technology that uses laser light to create high-resolution 3D maps of the environment. In robotics, LiDAR is used to build a 3D representation of the surroundings, allowing robots to navigate, map, and interact with their environment.

• **Applications in Robotics**: LiDAR is widely used in robotics applications such as autonomous vehicles, drone

### Prompt Constraint Experiment

The same LiDAR question was tested with three different output constraints: short, bullet-point, and technical formats.

**Observation:** The system prompt influenced the response style and structure. The bullet prompt produced a clear three-part answer, while the technical prompt produced a more detailed explanation using technical terminology. However, the model did not always follow strict constraints such as an exact sentence count.

**Key Learning:** Prompt constraints can guide LLM output format and level of detail, but they may not always be followed perfectly.

In [7]:
import ollama

combined_prompt = """
You are a robotics engineering tutor. Give clear, technically accurate answers for undergraduate students.

Follow the style shown in these examples.

Example 1:
Question: What is LiDAR?
Answer: LiDAR uses laser pulses to measure distances and create a representation of the environment.

Example 2:
Question: What is an ultrasonic sensor?
Answer: An ultrasonic sensor uses high-frequency sound waves to measure distance and detect nearby objects.

Now answer:
Question: What is a camera sensor in robotics?

Keep the answer under 40 words.
"""

response = ollama.chat(
    model="llama3.2:3b",
    messages=[
        {
            "role": "user",
            "content": combined_prompt
        }
    ]
)

print(response["message"]["content"])

A camera sensor in robotics is a type of image sensor that converts light into electrical signals, capturing visual data from the environment. It is commonly used in applications such as computer vision, object recognition, and autonomous navigation.


### Combined Prompting Experiment

A role-based instruction, few-shot examples, and an output-length constraint were combined in a single prompt.

**Observation:** The model followed the demonstrated answer style and produced a concise, technically relevant response. Combining multiple prompting techniques can provide better control over the content, style, and length of the generated response.

**Key Learning:** System instructions and few-shot examples can be combined to guide an LLM toward more consistent and controlled outputs.

In [8]:
import ollama

questions = [
    "What is ROS in robotics?",
    "Explain how a PID controller works.",
    "What is LiDAR and why is it useful for robots?"
]

models = ["llama3.2:3b", "qwen2.5:3b"]

for question in questions:
    print("\n" + "=" * 70)
    print("QUESTION:", question)
    print("=" * 70)

    for model in models:
        response = ollama.chat(
            model=model,
            messages=[
                {
                    "role": "system",
                    "content": "You are a robotics engineering tutor. Give clear and technically accurate answers for undergraduate students."
                },
                {
                    "role": "user",
                    "content": question
                }
            ]
        )

        print(f"\n--- {model} ---")
        print(response["message"]["content"])


QUESTION: What is ROS in robotics?

--- llama3.2:3b ---
ROS (Robot Operating System) is an open-source software framework that provides a common set of libraries and tools for building robot applications. It was originally developed by the Stanford Artificial Intelligence Lab (SAIL) and is now maintained by the Open Source Robotics Foundation (OSRF).

ROS provides a set of components that enable robot developers to:

1. **Communicate**: ROS allows robots to exchange data with other devices, such as sensors, actuators, and other computers.
2. **Coordinate**: ROS provides a framework for coordinating the activities of multiple robots, such as synchronization of timing and data sharing.
3. **Visualize**: ROS provides tools for visualizing the behavior of robots, such as 3D visualization and mapping.
4. **Program**: ROS provides a set of programming languages and tools, such as C++, Python, and Lua, that enable developers to write robot applications.

Some key features of ROS include:

* 

In [10]:
answer = response["message"]["content"]
print(f"\n{model}:")
print(answer)


qwen2.5:3b:
LiDAR (Light Detection and Ranging) uses laser light to measure distances, creating detailed 3D maps. It's crucial for robots in navigation and mapping environments with poor visibility.


In [11]:
import ollama

question = "What is ROS in robotics?"

for model in ["llama3.2:3b", "qwen2.5:3b"]:
    response = ollama.chat(
        model=model,
        messages=[
            {
                "role": "system",
                "content": "You are a robotics tutor. Answer in 60 words or less."
            },
            {
                "role": "user",
                "content": question
            }
        ]
    )

    print(f"\n===== {model} =====")
    print(response["message"]["content"])


===== llama3.2:3b =====
ROS (Robot Operating System) is an open-source software framework that enables the development of robot applications. It provides a set of tools, libraries, and protocols for building and maintaining robotic systems, including sensors, actuators, and autonomous navigation. ROS promotes modularity, reusability, and ease of maintenance, making it a popular choice among robotics developers.

===== qwen2.5:3b =====
ROS (Robot Operating System) is an open-source software framework for robotics, providing tools to develop robot applications.


In [12]:
import ollama

question = "Explain how a PID controller works."

for model in ["llama3.2:3b", "qwen2.5:3b"]:
    response = ollama.chat(
        model=model,
        messages=[
            {
                "role": "system",
                "content": "You are a robotics tutor. Answer in 60 words or less."
            },
            {
                "role": "user",
                "content": question
            }
        ]
    )

    print(f"\n===== {model} =====")
    print(response["message"]["content"])


===== llama3.2:3b =====
A PID (Proportional-Integral-Derivative) controller adjusts the output of a system to achieve a desired setpoint. It works by:

1. Proportional (P) component: Adjusts output based on the current error.
2. Integral (I) component: Adds credit for steady-state error over time.
3. Derivative (D) component: Compensates for rate of change of error.

The controller continuously calculates the error and adjusts the output accordingly.

===== qwen2.5:3b =====
A PID controller adjusts control pressure/speed based on error (setpoint - measurement), using Proportional, Integral, and Derivative terms to minimize error over time.


In [13]:
import ollama

response = ollama.chat(
    model="llama3.2:3b",
    messages=[
        {
            "role": "system",
            "content": "You are a robotics tutor. Answer in 60 words or less."
        },
        {
            "role": "user",
            "content": "What is LiDAR and why is it useful for robots?"
        }
    ]
)

print(response["message"]["content"])

LiDAR (Light Detection and Ranging) is a remote sensing technology that uses laser light to create high-resolution 3D maps of the environment. It's useful for robots as it enables accurate obstacle detection, navigation, and mapping, allowing them to navigate complex environments and avoid collisions. LiDAR is commonly used in autonomous vehicles, drones, and industrial robots.


### Llama 3.2 vs Qwen 2.5 Comparison

The same three robotics questions were tested using Llama 3.2:3b and Qwen 2.5:3b with the same system prompt.

**Observations:**
- Llama 3.2:3b generally produced more detailed and explanatory responses.
- Qwen 2.5:3b generally produced shorter and more direct responses.
- Both models provided relevant answers to the robotics questions.
- The response style and level of detail differed even when the same prompt was used.

**Key Learning:** Comparing multiple local LLMs with identical prompts helps evaluate differences in response style, detail, and usefulness for a specific task.